tutelary/models.py

```diff
@@ -197,25 +188,26 @@ def by_policies_and_roles(self, policies_roles):
                 canonpols.append((pr, vars, None))
 
         # Try to find an existing permission set using all the same
-        # policies and variable assignments.  First step is to find a
-        # list of permission set IDs using all the same policies as
-        # appear in the supplied policy list.
-        for psetid in _policy_psets(canonpols):
+        # policies and variable assignments.
+        matching_psets = _policy_psets(canonpols).annotate(
+            num_pis=models.Count('policyinstance')
+        ).filter(
+            # Same number of policy instances
+            num_pis=len(canonpols)
+        )
+
+        for policy, variables, role in canonpols:
             # Now, for each permission set candidate, check whether
             # the policy instance rows correspond exactly to the input
             # policy+variable assignment list.
-            pset = self.get(id=psetid)
-            pis = PolicyInstance.objects.filter(pset=pset)
-            if len(pis) == len(canonpols):
-                same = True
-                for pi, canonpol in zip(pis, canonpols):
-                    if (pi.variables != canonpol[1] or
-                       (canonpol[2] is not None and pi.role != canonpol[0]) or
-                       (canonpol[2] is None and pi.policy != canonpol[0])):
```

Comment:
For the record, I don't believe this ever worked. I believe this should be (note the change from `canonpol[0]` on the last two lines):

```py
if (pi.variables != canonpol[1] or
    (canonpol[2] is not None and pi.role != canonpol[2]) or
    (canonpol[0] is None and pi.policy != canonpol[0])):
```

Reply:
I agree with your assessment that the original code does not work as intended. But this implies that tutelary has a bug and therefore we need a new unit test to detect this bug and confirm that your changes fixes the bug.

Additionally, I wonder if this bug is the cause of some of our weird permission errors in the platform. What do you think?

Reply:
While this probably will not instil confidence in this PR, I will admit that I don't totally grasp how all of this code is used (I merely just tried to replace logic with querysets that would produce the same results).  So, let's walk through this together: This code block aims to find policy instances that exactly matches the variables, role, and policy provided by the "canonical input policy list". Because of the bug, `same = False` will always be the case.  This means that we then create a new `PermissionSet` instance and and a number of `PolicyInstance` instances.

This is all part of the `PermissionSetManager.objects.by_policies_and_roles` method. I can only find this method being called in the [`assign_user_policies` function](https://github.com/Cadasta/django-tutelary/blob/041139d3a7174d6e9340418afa10b2a3ee6dc3bc/tutelary/models.py#L323-L330).  Aside from tests, I can only find that function called in one place: the [`accounts.load.run` function](https://github.com/Cadasta/cadasta-platform/blob/a880b13661ebb62229a4c6a34c5e141d27ae86a9/cadasta/accounts/load.py#L35), which only runs from the [loadpolicies management command](https://github.com/Cadasta/cadasta-platform/blob/a880b13661ebb62229a4c6a34c5e141d27ae86a9/cadasta/accounts/management/commands/loadpolicies.py).  So, all of this means that the function really only affects new DBs and thus the bug probably wasn't a serious issue for anyone.

I didn't really look at _when_ this code gets used, moreso I was just optimizing where I could see opportunities.  Seems like the bug wasn't a big deal.

Do you still feel I should write a test for this?  I feel like I probably should, but life would go on without one.

Reply:
That was my hunch as well. I think we would just end up creating even new permission set objects even if there were existing matching ones already. Note though that `User.assign_policies` is effectively an [alias](https://github.com/Cadasta/django-tutelary/blob/4131599a90cb853a9d9ca88d33a22901904f1e8d/tutelary/apps.py) to `assign_user_policies`, so you should also be looking for calls to that method as well.

I definitely know that we create new permission sets when a user gets assigned/reassigned/removed roles (org or project). So I think the only problem this bug causes is that we may end up with a lot of stale permission set objects that are not used. But I don't think this would cause the user to end up with the wrong permissions.

OK. I think we don't need a test for this tiny bug. :D